atlasdb-commons/src/test/java/com/palantir/util/IndexEncodingUtilsTest.java

```diff
@@ -0,0 +1,207 @@
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class IndexEncodingUtilsTest {
+
+    @Parameterized.Parameters(name = "checksumType={0}")
+    public static Iterable<ChecksumType> data() {
+        return Arrays.asList(ChecksumType.values());
+    }
+
+    private static final Map<DH<String>, Long> VALUES =
+            ImmutableMap.of(dh("key1"), 42L, dh("key2"), 1337L, dh("anotherKey"), -10L);
+    private static final Set<DH<String>> KEYS = ImmutableSet.of(dh("key1"), dh("key2"), dh("anotherKey"));
+    private static final Map<Integer, Long> INDEX_ENCODED_VALUES = ImmutableMap.of(1, 1337L, 2, -10L, 0, 42L);
+    private static final Function<Long, String> VALUE_MAPPER = value -> Long.toString(value - 10);
+    private static final Function<String, Long> REVERSE_MAPPER = value -> Long.parseLong(value) + 10;
+
+    private final ChecksumType checksumType;
+    private IndexEncodingResult<DH<String>, Long> encoded;
+
+    public IndexEncodingUtilsTest(ChecksumType checksumType) {
+        this.checksumType = checksumType;
+    }
+
+    @Before
+    public void setup() {
+        encoded = IndexEncodingUtils.encode(KEYS, VALUES, Function.identity(), checksumType);
+    }
+
+    @Test
+    public void canLookupValidChecksumType() {
+        assertThat(ChecksumType.valueOf(checksumType.getId())).isEqualTo(checksumType);
+    }
+
+    @Test
+    public void cannotLookupUnknownChecksumType() {
+        assertThatException()
+                .isThrownBy(() -> ChecksumType.valueOf(-1))
+                .isInstanceOf(SafeIllegalArgumentException.class)
+                .withMessage(
+                        SafeExceptions.renderMessage("Unknown checksum type ID", SafeArg.of("checksumTypeId", -1)));
+    }
+
+    @Test
+    public void canEncodeSimpleData() {
+        assertThat(encoded.indexToValue()).isEqualTo(INDEX_ENCODED_VALUES);
+    }
+
+    @Test
+    public void canDecodeSimpleData() {
+        assertThat(IndexEncodingUtils.decode(encoded, Function.identity())).containsExactlyInAnyOrderEntriesOf(VALUES);
+    }
+
+    @Test
+    public void canEncodeWithCustomValueMapper() {
+        Map<Integer, String> expectedValues = KeyedStream.ofEntries(INDEX_ENCODED_VALUES.entrySet().stream())
+                .map(VALUE_MAPPER)
+                .collectToMap();
+        assertThat(IndexEncodingUtils.encode(KEYS, VALUES, VALUE_MAPPER, checksumType)
+                        .indexToValue())
+                .containsExactlyInAnyOrderEntriesOf(expectedValues);
+    }
+
+    @Test
+    public void canDecodeWithCustomValueMapper() {
+        assertThat(IndexEncodingUtils.decode(
+                        IndexEncodingUtils.encode(KEYS, VALUES, VALUE_MAPPER, checksumType), REVERSE_MAPPER))
+                .containsExactlyInAnyOrderEntriesOf(VALUES);
+    }
+
+    @Test
+    public void encodeFailsForUnknownKeysInValueMap() {
+        assertThatException()
```

Comment:
Let's use assertThatLoggableExceptionThrownBy, but you'll need to add the relevant test dependency.